framework/include/base/DependencyResolverInterface.h

```diff
@@ -15,49 +15,100 @@
 #ifndef DEPENDENCYRESOLVERINTERFACE_H
 #define DEPENDENCYRESOLVERINTERFACE_H
 
+// STL includes
 #include <string>
 #include <set>
 #include <iostream>
 #include <algorithm>
 
+// MOOSE includes
 #include "DependencyResolver.h"
 
+/**
+ * Interface for sorting dependent vectors of objects.
+ */
 class DependencyResolverInterface
 {
 public:
+  /**
+   * Constructor.
+   */
   DependencyResolverInterface() {}
 
+  /**
+   * Return a set containing the names of items requested by the object.
+   */
   virtual const std::set<std::string> & getRequestedItems() = 0;
+
+  /**
+   * Return a set containing the names of items owned by the object.
+   */
   virtual const std::set<std::string> & getSuppliedItems() = 0;
 
-  template<typename Iter>
+  /**
+   * Given a vector, sort using the getRequested/SuppliedItems sets.
+   */
+  template<typename T>
+  static
+  void sort(typename std::vector<T> & vector);
+
+  /**
+   * A helper method for cyclic errors.
+   * @see MaterialWarehouse
+   */
+  template<typename T>
   static
-  void sort(Iter start, Iter end)
+  void cyclicDependencyError(CyclicDependencyException<T> & e, const std::string & header);
+};
+
+
+template<typename T>
+void
+DependencyResolverInterface::sort(typename std::vector<T> & vector)
```

Comment:
Because of the vector MooseSharedPointers versus vector of pointers. When passing regular pointers an implicit cast occurs in the template methods, this is not possible with the shared pointers.